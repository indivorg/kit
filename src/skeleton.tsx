import { keyframes } from '@emotion/react';
import React from 'react';
import { Box, BoxProps } from 'theme-ui';

export interface SkeletonProps extends BoxProps {
  /** `circular` | `rectangular` | `text` */
  shape?: 'circular' | 'rectangular' | 'text';

  /**
   * Use to set size of circular shape.
   *
   * @default 1.2rem
   */
  size?: string | number;
}

const pulseKeyframe = keyframes({
  '0%': {
    opacity: 1,
  },
  '50%': {
    opacity: 0.4,
  },
  '100%': {
    opacity: 1,
  },
});

export const Skeleton = React.forwardRef<HTMLDivElement, SkeletonProps>(
  ({ shape = 'text', size = '1.2em', ...props }, ref) => (
    <Box
      ref={ref}
      variant="primary"
      {...props}
      // @ts-expect-error themeKey is a private prop
      __themeKey="skeleton"
      __css={{
        backgroundColor: 'muted',
        height: shape === 'circular' && size,
        width: shape === 'circular' && size,
        borderRadius: shape === 'circular' ? '50%' : 4,
        transformOrigin: shape === 'text' && '0 55%',
        transform: shape === 'text' && 'scale(1, 0.60)',
        animation: `${pulseKeyframe} 1.5s ease-in-out 0.5s infinite`,
        '&:empty:before': {
          content: '"\\00a0"',
        },
      }}
    />
  ),
);
