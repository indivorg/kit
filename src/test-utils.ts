import renderer from 'react-test-renderer';

export const renderJSON = (
  el: Parameters<typeof renderer.create>[0],
): renderer.ReactTestRendererJSON | null => {
  const json = renderer.create(el).toJSON();
  if (Array.isArray(json)) {
    return json[0];
  }

  return json;
};
