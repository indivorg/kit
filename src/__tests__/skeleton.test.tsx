/**
 * @jest-environment jsdom
 */
import { matchers } from '@emotion/jest';
import React from 'react';
import { Skeleton } from '../skeleton';
import { renderJSON } from '../test-utils';

expect.extend(matchers);

describe('skeleton', () => {
  it.each(['text', 'circular', 'rectangular'] as const)(
    'should render %s',
    shape => {
      const json = renderJSON(<Skeleton shape={shape} />);
      expect(json).toMatchSnapshot();
    },
  );

  it('should always have muted background', () => {
    const circle = renderJSON(<Skeleton shape="circular" />);
    expect(circle).toHaveStyleRule('background-color', 'muted');
  });
});
