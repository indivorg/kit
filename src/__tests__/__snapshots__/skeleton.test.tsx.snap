// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`skeleton should render circular 1`] = `
<div
  className="css-tg8bh6"
/>
`;

exports[`skeleton should render rectangular 1`] = `
<div
  className="css-1v4u1j6"
/>
`;

exports[`skeleton should render text 1`] = `
<div
  className="css-8yc0al"
/>
`;
